Fix some data races on shutdown of cilium-health

joestringer opened the pull request:
This series also does some cleanup / simplification, removing a couple of goroutines from this code.

`Reported-by: André Martins <andre@cilium.io>`

Remaining tasks:
* [x] Fix referencing `p.nodes` in `runHTTPProbe()`
* [x] Ensure that `setNodes()` only makes the nodes list available after setting up the map.
* [x] Consolidate logic for choosing IPs to probe via ICMP/HTTP

aanm reviewed (dismissed):
Just want to know if you are aware that we can panic if accessing a non-existing key
```
prober.results[ipString(addr.String())].Icmp  
p.results[ipString(elem.IP)].HTTP
```

if `prober.results[ipString(addr.String())]` or `prober.results[ipString(elem.IP)]` or doesn't exist cilium panics.

joestringer committed:
cilium-health: Remove server.nodes

This was originally the place where the nodes list was stored, but it's
unnecessary in here since the prober holds the list of nodes now. Remove
this unused reference to the map.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Fix another data race

The self status was being passed around by reference without locking, so
building cilium-health with '-race' then killing the process would
reveal this data race in the logs. Fix it since it's trivial.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Remove unnecessary goroutine

Prior to this commit, we had an additional goroutine that wrapped the
pinger/prober that waited until they both complete to print a "finished"
message. This goroutine was in charge of building the 'prober' object,
which would be accessed from the main thread to halt the thread via
Stop(). If the logic actually followed to loop around and recreate the
prober, then this could race with server shutdown from the main thread
and have undefined behaviour. Removing the goroutine removes this data
race and simplifies the code.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
prober: Make prober.Done() more consistent

Prior to this commit, when the main thread (doing API serving) starts to
shut down, it would call prober.Stop(), which would launch another
goroutine that attempted to proxy the pinger and prober's finish result
into a freshly returned socket. The main thread would wait on the
returned socket. However, multiple threads ended up racing to read from
the 'p.done' socket.

This commit splits it out so that the p.Pinger().Stop() function is
called, after which the pinger is finished; then it closes the p.stop
channel, signalling to the HTTP prober goroutine to finish. Once that
goroutine finishes, it signals the new p.proberExited channel, which is
listened from the thread running the p.Stop(). Finally, it can signal
the Done() socket and return.

This fixes some data race issues on shutdown, and simplifies the
goroutine model in cilium-health.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
prober: Fix access of nodes map from runHTTPProbe

Previously, a pointer to the map would be copied, then the contents of
the map would be iterated across without any locking or synchronization.
This would race with the pinger thread running setNodes() to update the
nodes list. Fix this race by making a local copy of nodes that shuold be
pinged first, then iterating across that set.

Signed-off-by: Joe Stringer <joestringernz@gmail.com>

joestringer committed:
prober: Consolidate IP selection for probing

Previously, the ICMP and HTTP probers could be inconsistent in the set
of hosts that they selected to ping, because the ICMP ping would skip
garbage IPs (because they had to be resolved first), whereas the HTTP
prober would attempt to ping them. Consolidate the validation of the
IPs so that these paths are consistent.

Signed-off-by: Joe Stringer <joestringernz@gmail.com>

aanm commented on pkg/health/server/prober.go:
```diff
@@ -228,45 +238,42 @@ func (p *prober) httpProbe(node string, ip string, port int) *models.Connectivit
 		}
+
+		p.Lock()
+		p.results[ipString(ip.String())].HTTP = status.HTTP
```
If `p.results[ipString(ip.String())]` returns nil, because it can return nil if `ipString(ip.String())` doesn't exist on the map, you get a panic for trying to access a nil pointer.

joestringer replied:
This code should only be called for an IP that is in the nodes list, and `setNodes()` should ensure that entries are only there if an entry in `p.results` is created for it as well. That said, I may have missed something so if you're still concerned about it, I can always add the extra check.

jrfastab reviewed (approved):
LGTM
